### DISCUSSION TITLE
Proposal: Anonymous inner class

### DISCUSSION THREAD

**Discussion by pengweiqhca** [dotnet/csharplang#4148] [Language Ideas] [answered]
# Anonymous inner class

* [x] Proposed
* [ ] Prototype: Not Started
* [ ] Implementation: Not Started
* [ ] Specification: Not Started

## Summary
A type is only used once, and don't want to be named.
Java syntax:
``` java
public class TestClass
{
    private void TestMethod()
    {
        MyInterface localVar = new MyInterface()
        {
            public void method1()
            {
                someCode();
            }
            public void method2(int i, boolean b)
            {
                someCode();
            }
        };
    }
}
```
The corresponding C# implementation:
``` C#
public class TestClass
{
    private void TestMethod()
    {
        MyInterface localVar = new MyInterfaceAnonymousInnerClass(this);
    }

    private class MyInterfaceAnonymousInnerClass : MyInterface
    {
        private readonly TestClass outerInstance;

        public MyInterfaceAnonymousInnerClass(TestClass outerInstance)
        {
            this.outerInstance = outerInstance;
        }

        public void method1()
        {
            someCode();
        }
        public void method2(int i, bool b)
        {
            someCode();
        }
    }
}
```

**Comment by JustNrik** [comment]
C# already supports anonymous types, but they can't define methods or implement interfaces, Java allow this because they don't have delegates, so this is their way to work around them, other than that, I don't see where this would be useful

**Reply by pengweiqhca** [reply]
But not always can use delegate.

**Reply by theunrepentantgeek** [reply]
@RUSshy there’s a growing problem of people throwing tantrums because they disagree with the decisions made by the LDM on the future of C#. I suggest it’s unwise of you to fan those flames by encouraging poor behaviour.

**Reply by HaloFour** [reply]
You're welcome to oppose the decisions and the reasons for those decisions as made by the LDM.  But without bringing new data to the equation the likelihood that such opposition will have any impact on the decision making process is negligible.  The _vast_ majority of proposals for language changes are not going to be accepted into the language.

**Reply by CyrusNajmabadi** [reply]
> sometimes the LDM needs to understand that the world have different opinions

And sometimes those with different opinions need to understand that those opinions were heard but a decision was still made that goes against their perspective.

**Reply by Joe4evr** [reply]
> and you can't even do polymorphism with delegates

But who knows what the future will bring? [There's been some interest in having polymorphic delegates by the ASP.NET team](https://github.com/dotnet/runtime/issues/8362#issuecomment-725723396).

**Comment by YairHalberstadt** [comment, marked as answer]
Duplicate of https://github.com/dotnet/csharplang/discussions/738, https://github.com/dotnet/csharplang/discussions/1542, https://github.com/dotnet/csharplang/discussions/2298
